README.md
# baldor
[![Build Status](https://travis-ci.org/crigroup/baldor.svg?branch=master)](https://travis-ci.org/crigroup/baldor) [![Coverage Status](https://coveralls.io/repos/github/crigroup/baldor/badge.svg)](https://coveralls.io/github/crigroup/baldor)

ROS (Python) package to work with Homogeneous Transformation Matrices, Quaternions, Euler angles, axis-angle rotations.

This package has been developed by [CRI Group](http://www.ntu.edu.sg/home/cuong/),
[Nanyang Technological University, Singapore](http://www.ntu.edu.sg).

This package is based on:
- [transformations.py](http://www.lfd.uci.edu/~gohlke/code/transformations.py.html) by Christoph Gohlke
- [transforms3d](http://matthew-brett.github.io/transforms3d) by Matthew Brett.

### Maintainer
* [Francisco Suárez Ruiz](http://fsuarez6.github.io)

### Documentation
* Throughout the various files in this repository.
* Website: http://wiki.ros.org/baldor

## ROS Buildfarm

ROS Distro | Source | Debian | Release Status
---------- | ------ | ------ | --------------
kinetic | [![Build Status](http://build.ros.org/buildStatus/icon?job=Ksrc_uX__baldor__ubuntu_xenial__source)](http://build.ros.org/job/Ksrc_uX__baldor__ubuntu_xenial__source/) | [![Build Status](http://build.ros.org/buildStatus/icon?job=Kbin_uX64__baldor__ubuntu_xenial_amd64__binary)](http://build.ros.org/job/Kbin_uX64__baldor__ubuntu_xenial_amd64__binary/) | <a href="http://repositories.ros.org/status_page/ros_kinetic_default.html?q=baldor">Status</a>
melodic | [![Build Status](http://build.ros.org/buildStatus/icon?job=Msrc_uB__baldor__ubuntu_bionic__source)](http://build.ros.org/view/Msrc_uB/job/Msrc_uB__baldor__ubuntu_bionic__source/) | [![Build Status](http://build.ros.org/buildStatus/icon?job=Mbin_uB64__baldor__ubuntu_bionic_amd64__binary)](http://build.ros.org/view/Mbin_uB64/job/Mbin_uB64__baldor__ubuntu_bionic_amd64__binary/) | <a href="http://repositories.ros.org/status_page/ros_melodic_default.html?q=baldor">Status</a>
noetic  |  |  | <a href="http://repositories.ros.org/status_page/ros_noetic_default.html?q=baldor">Status</a>

Check in the *release status* which versions of the package are in **building**, **ros-shadow-fixed**
(tagged as 'testing'), and **ros** (tagged as 'main').

Approximately every two weeks, the ROS platform manager manually synchronizes
the contents of **ros-shadow-fixed** into **ros** (the public repository).


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package baldor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.1.3 (2020-10-25)
------------------
* Add support for ROS noetic and clean up
* Contributors: fsuarez6

0.1.2 (2017-12-30)
------------------
* Add function to estimate the transformation between two axes (#3)
* Contributors: Francisco

0.1.1 (2017-11-03)
------------------
* Add ROS buildfarm status
* Add coveralls support
* Contributors: Francisco

0.1.0 (2017-10-24)
------------------
* Initial release
* Contributors: fsuarez6


tests/test_euler.py
#! /usr/bin/env python
import unittest
import numpy as np
# Tested package
import baldor as br


class TestModule(unittest.TestCase):
    def test_axes_sequences(self):
        for axes in br._AXES2TUPLE.keys():
            ai, aj, ak = np.random.sample(3)*2*np.pi
            # Axis-angle
            axis, angle = br.euler.to_axis_angle(ai, aj, ak, axes)
            # Transform
            T = br.euler.to_transform(ai, aj, ak, axes)

    def test_to_axis_angle(self):
        axis, angle = br.euler.to_axis_angle(0, 1.5, 0, 'szyx')
        np.testing.assert_allclose(axis, [0, 1, 0])
        np.testing.assert_almost_equal(angle, 1.5)

    def test_to_quaternion(self):
        q = br.euler.to_quaternion(1, 2, 3, 'ryxz')
        expected = [0.43595284, 0.31062245, -0.71828702, 0.44443511]
        np.testing.assert_allclose(q, expected)
        q = br.euler.to_quaternion(1, 2, 3, (2, 0, 0, 1))
        # Test parity
        q = br.euler.to_quaternion(1, 2, 3, 'rzyz')
        expected = [-0.2248451,  0.70807342,  0.45464871,  0.4912955]
        np.testing.assert_allclose(q, expected)

    def test_to_transform(self):
        T = br.euler.to_transform(1, 2, 3, 'syxz')
        np.testing.assert_almost_equal(np.sum(T[0]), -1.34786452)
        T = br.euler.to_transform(1, 2, 3, (0, 1, 0, 1))
        np.testing.assert_almost_equal(np.sum(T[0]), -0.383436184)


tests/test_vector.py
#! /usr/bin/env python
import unittest
import numpy as np
# Tested package
import baldor as br


class TestModule(unittest.TestCase):
    def test_unit(self):
        v = np.random.sample(3)
        u = br.vector.unit(v)
        np.testing.assert_allclose(np.linalg.norm(u), 1, atol=1e-8)

    def test_norm(self):
        v = np.random.sample(3)
        np.testing.assert_allclose(
            br.vector.norm(v), np.linalg.norm(v), atol=1e-8)

    def test_perpendicular(self):
        u = np.random.sample(3)
        v = br.vector.perpendicular(u)
        np.testing.assert_allclose(np.dot(u, v), 0, atol=1e-8)

        # Test zero vector
        def zero_vector_test():
            br.vector.perpendicular(np.zeros(3))

        self.assertRaises(ValueError, zero_vector_test)
        # Test Z_AXIS case
        u1 = br.Z_AXIS
        v1 = br.vector.perpendicular(u1)
        np.testing.assert_allclose(np.dot(u1, v1), 0, atol=1e-8)

    def test_skew(self):
        vector = np.random.sample(3)
        R = br.vector.skew(vector)
        np.testing.assert_allclose(-R, R.T)

    def test_transform_between_vectors(self):
        u0 = br.vector.unit(np.random.sample(3))
        q = br.quaternion.random()
        T0 = br.quaternion.to_transform(q)
        v0 = br.vector.unit(np.dot(T0, np.hstack((u0, 1)))[:3])
        T1 = br.vector.transform_between_vectors(u0, v0)
        v1 = br.vector.unit(np.dot(T1, np.hstack((u0, 1)))[:3])
        np.testing.assert_allclose(v0, v1, atol=1e-8)
        # Test same vector
        T2 = br.vector.transform_between_vectors(u0, u0)
        np.testing.assert_allclose(T2, np.eye(4), atol=1e-8)


tests/test_transform.py
#! /usr/bin/env python
import unittest
import numpy as np
# Tested package
import baldor as br


class TestModule(unittest.TestCase):
    def test_are_equal(self):
        T0 = np.diag([1, 1, 1, 1])
        T1 = np.eye(4)
        self.assertTrue(br.quaternion.are_equal(T0, T1))
        T2 = br.quaternion.to_transform([1, 0, 0, 0])
        T3 = br.quaternion.to_transform([-1, 0, 0, 0])
        self.assertTrue(br.quaternion.are_equal(T2, T3))

    def test_between_axes(self):
        # Random axis / angle
        np.random.seed(123)
        axis = br.vector.unit(np.random.randn(3))
        angle = np.deg2rad(45)
        transform = br.axis_angle.to_transform(axis, angle)
        newaxis = np.dot(transform[:3, :3], br.Z_AXIS)
        est_transform = br.transform.between_axes(br.Z_AXIS, newaxis)
        np.testing.assert_allclose(transform[:3, 2], est_transform[:3, 2])
        # Edge case 1
        newaxis = -br.Z_AXIS
        transform = br.transform.between_axes(br.Z_AXIS, newaxis)
        _, angle, _ = br.transform.to_axis_angle(transform)
        np.testing.assert_allclose(angle, np.pi)
        # Edge case 2
        newaxis = br.Z_AXIS
        transform = br.transform.between_axes(-br.Z_AXIS, newaxis)
        _, angle, _ = br.transform.to_axis_angle(transform)
        np.testing.assert_allclose(angle, np.pi)

    def test_inverse(self):
        q = br.quaternion.random()
        T = br.quaternion.to_transform(q)
        Tinv = br.transform.inverse(T)
        np.testing.assert_allclose(
            np.dot(T, Tinv), np.eye(4), rtol=1e-5, atol=1e-8)
        np.testing.assert_allclose(
            Tinv, np.linalg.inv(T), rtol=1e-5, atol=1e-8)

    def test_random(self):
        T = br.transform.random()
        Tinv = br.transform.inverse(T)
        np.testing.assert_allclose(np.dot(T, Tinv), np.eye(4), atol=1e-8)

    def test_to_axis_angle(self):
        axis = np.random.sample(3) - 0.5
        angle = (np.random.sample(1) - 0.5) * (2*np.pi)
        point = np.random.sample(3) - 0.5
        T0 = br.axis_angle.to_transform(axis, angle, point)
        axis, angle, point = br.transform.to_axis_angle(T0)
        T1 = br.axis_angle.to_transform(axis, angle, point)
        self.assertTrue(br.transform.are_equal(T0, T1))
        # The eigenvector of the rotation matrix corresponding to eigenvalue 1 is
        # the axis of rotation

        def rotation_eigenvalue_test():
            T2 = np.eye(4)
            T2[:3, :3] *= 2
            br.transform.to_axis_angle(T2)
        self.assertRaises(ValueError, rotation_eigenvalue_test)
        # Test simple axes
        T = br.axis_angle.to_transform(br.X_AXIS, angle)
        v0, angle, _ = br.transform.to_axis_angle(T)
        np.testing.assert_allclose(v0, br.X_AXIS, atol=1e-8)
        T = br.axis_angle.to_transform(br.Y_AXIS, angle)
        v1, angle, _ = br.transform.to_axis_angle(T)
        np.testing.assert_allclose(v1, br.Y_AXIS, atol=1e-8)
        T = br.axis_angle.to_transform(br.Z_AXIS, angle)
        v2, angle, _ = br.transform.to_axis_angle(T)
        np.testing.assert_allclose(v2, br.Z_AXIS, atol=1e-8)

    def test_to_dual_quaternion(self):
        qr0 = br.quaternion.random()
        qt0 = br.quaternion.random()
        # Check transform consistency
        T0 = br.quaternion.dual_to_transform(qr0, qt0)
        Tinv = br.transform.inverse(T0)
        np.testing.assert_allclose(
            np.dot(T0, Tinv), np.eye(4), rtol=1e-5, atol=1e-8)
        # Convert back
        qr1, qt1 = br.transform.to_dual_quaternion(T0)
        T1 = br.quaternion.dual_to_transform(qr1, qt1)
        # T0 and T1 must be equal
        self.assertTrue(br.transform.are_equal(T0, T1))

    def test_to_euler(self):
        # Test all tuples few times
        for axes in br._AXES2TUPLE.keys():
            ai, aj, ak = np.random.sample(3)*2*np.pi
            T0 = br.euler.to_transform(ai, aj, ak, axes)
            al, be, ga = br.transform.to_euler(T0, axes)
            T1 = br.euler.to_transform(al, be, ga, axes)
            np.testing.assert_allclose(T0, T1)

    def test_to_quaternion(self):
        # Identity test
        q = br.transform.to_quaternion(np.identity(4), isprecise=True)
        np.testing.assert_allclose(q, [1, 0, 0, 0])
        # Simple rotation test
        q = br.transform.to_quaternion(np.diag([1, -1, -1, 1]))
        self.assertTrue(np.allclose(q, [0, 1, 0, 0]) or
                        np.allclose(q, [0, -1, 0, 0]))
        # Dummy test
        T0 = br.axis_angle.to_transform((1, 2, 3), 0.123)
        q = br.transform.to_quaternion(T0, True)
        np.testing.assert_allclose(q, [0.9981095, 0.0164262, 0.0328524, 0.0492786],
                                   rtol=1e-05, atol=1e-8)
        # Permutation matrix
        T1 = [[0, 0, 1, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0, 1, 0, 1]]
        q1 = br.transform.to_quaternion(T1, True)
        # Round trip
        np.random.seed(123)   # Generate always the same transforms
        for _ in range(100):
            T0 = br.transform.random()
            T0[:3, 3] = 0
            isprecise = np.random.rand() > 0.5
            q0 = br.transform.to_quaternion(T0, isprecise)
            T1 = br.quaternion.to_transform(q0)
            self.assertTrue(br.transform.are_equal(T0, T1))


src/baldor/euler.py
#!/usr/bin/env python
"""
Generic Euler rotations
"""
import math
import numpy as np
# Local modules
import baldor as br


def to_axis_angle(ai, aj, ak, axes='sxyz'):
    """
    Return axis-angle rotation from Euler angles and axes sequence

    Parameters
    ----------
    ai: float
      First rotation angle (according to axes).
    aj: float
      Second rotation angle (according to axes).
    ak: float
      Third rotation angle (according to axes).
    axes: str, optional
      Axis specification; one of 24 axis sequences as string or encoded tuple

    Returns
    -------
    axis: array_like
      axis around which rotation occurs
    angle: float
      angle of rotation

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> axis, angle = br.euler.to_axis_angle(0, 1.5, 0, 'szyx')
    >>> np.allclose(axis, [0, 1, 0])
    True
    >>> angle
    1.5
    """
    T = to_transform(ai, aj, ak, axes)
    axis, angle, _ = br.transform.to_axis_angle(T)
    return axis, angle


def to_quaternion(ai, aj, ak, axes='sxyz'):
    """
    Returns a quaternion from Euler angles and axes sequence

    Parameters
    ----------
    ai: float
      First rotation angle (according to axes).
    aj: float
      Second rotation angle (according to axes).
    ak: float
      Third rotation angle (according to axes).
    axes: str, optional
      Axis specification; one of 24 axis sequences as string or encoded tuple

    Returns
    -------
    q: array_like
      Quaternion in w, x, y z (real, then vector) format

    Notes
    -----
    Quaternions :math:`w + ix + jy + kz` are represented as :math:`[w, x, y, z]`.

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> q = br.euler.to_quaternion(1, 2, 3, 'ryxz')
    >>> np.allclose(q, [0.435953, 0.310622, -0.718287, 0.444435])
    True
    """
    try:
        firstaxis, parity, repetition, frame = br._AXES2TUPLE[axes.lower()]
    except (AttributeError, KeyError):
        br._TUPLE2AXES[axes]  # validation
        firstaxis, parity, repetition, frame = axes

    i = firstaxis + 1
    j = br._NEXT_AXIS[i+parity-1] + 1
    k = br._NEXT_AXIS[i-parity] + 1

    if frame:
        ai, ak = ak, ai
    if parity:
        aj = -aj

    ai /= 2.0
    aj /= 2.0
    ak /= 2.0
    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    cc = ci*ck
    cs = ci*sk
    sc = si*ck
    ss = si*sk

    q = np.empty((4, ))
    if repetition:
        q[0] = cj*(cc - ss)
        q[i] = cj*(cs + sc)
        q[j] = sj*(cc + ss)
        q[k] = sj*(cs - sc)
    else:
        q[0] = cj*cc + sj*ss
        q[i] = cj*sc - sj*cs
        q[j] = cj*ss + sj*cc
        q[k] = cj*cs - sj*sc
    if parity:
        q[j] *= -1.0
    return q


def to_transform(ai, aj, ak, axes='sxyz'):
    """
    Return homogeneous transformation matrix from Euler angles and axes sequence.

    Parameters
    ----------
    ai: float
      First rotation angle (according to axes).
    aj: float
      Second rotation angle (according to axes).
    ak: float
      Third rotation angle (according to axes).
    axes: str, optional
      Axis specification; one of 24 axis sequences as string or encoded tuple

    Returns
    -------
    T: array_like
      Homogeneous transformation (4x4)

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> T = br.euler.to_transform(1, 2, 3, 'syxz')
    >>> np.allclose(np.sum(T[0]), -1.34786452)
    True
    >>> T = br.euler.to_transform(1, 2, 3, (0, 1, 0, 1))
    >>> np.allclose(np.sum(T[0]), -0.383436184)
    True
    """
    try:
        firstaxis, parity, repetition, frame = br._AXES2TUPLE[axes]
    except (AttributeError, KeyError):
        br._TUPLE2AXES[axes]  # validation
        firstaxis, parity, repetition, frame = axes

    i = firstaxis
    j = br._NEXT_AXIS[i+parity]
    k = br._NEXT_AXIS[i-parity+1]

    if frame:
        ai, ak = ak, ai
    if parity:
        ai, aj, ak = -ai, -aj, -ak

    si, sj, sk = math.sin(ai), math.sin(aj), math.sin(ak)
    ci, cj, ck = math.cos(ai), math.cos(aj), math.cos(ak)
    cc, cs = ci*ck, ci*sk
    sc, ss = si*ck, si*sk

    T = np.identity(4)
    if repetition:
        T[i, i] = cj
        T[i, j] = sj*si
        T[i, k] = sj*ci
        T[j, i] = sj*sk
        T[j, j] = -cj*ss+cc
        T[j, k] = -cj*cs-sc
        T[k, i] = -sj*ck
        T[k, j] = cj*sc+cs
        T[k, k] = cj*cc-ss
    else:
        T[i, i] = cj*ck
        T[i, j] = sj*sc-cs
        T[i, k] = sj*cc+ss
        T[j, i] = cj*sk
        T[j, j] = sj*ss+cc
        T[j, k] = sj*cs-sc
        T[k, i] = -sj
        T[k, j] = cj*si
        T[k, k] = cj*ci
    return T


src/baldor/vector.py
#!/usr/bin/env python
"""
Vector operations
"""
import math
import numpy as np
# Local modules
import baldor as br


def unit(vector):
    """
    Return vector divided by Euclidean (L2) norm

    Parameters
    ----------
    vector: array_like
      The input vector

    Returns
    -------
    unit : array_like
      Vector divided by L2 norm

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> v0 = np.random.random(3)
    >>> v1 = br.vector.unit(v0)
    >>> np.allclose(v1, v0 / np.linalg.norm(v0))
    True
    """
    v = np.asarray(vector).squeeze()
    return v / math.sqrt((v**2).sum())


def norm(vector):
    """
    Return vector Euclidaan (L2) norm

    Parameters
    ----------
    vector: array_like
      The input vector

    Returns
    -------
    norm: float
      The computed norm

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> v = np.random.random(3)
    >>> n = br.vector.norm(v)
    >>> numpy.allclose(n, np.linalg.norm(v))
    True
    """
    return math.sqrt((np.asarray(vector)**2).sum())


def perpendicular(vector):
    """
    Find an arbitrary perpendicular vector

    Parameters
    ----------
    vector: array_like
      The input vector

    Returns
    -------
    result: array_like
      The perpendicular vector
    """
    if np.allclose(vector, np.zeros(3)):
        # vector is [0, 0, 0]
        raise ValueError('Input vector cannot be a zero vector')
    u = unit(vector)
    if np.allclose(u[:2], np.zeros(2)):
        return br.Y_AXIS
    result = np.array([-u[1], u[0], 0], dtype=np.float64)
    return result


def skew(vector):
    """
    Returns the 3x3 skew matrix of the input vector.

    The skew matrix is a square matrix `R` whose transpose is also its negative;
    that is, it satisfies the condition :math:`-R = R^T`.

    Parameters
    ----------
    vector: array_like
      The input array

    Returns
    -------
    R: array_like
      The resulting 3x3 skew matrix
    """
    skv = np.roll(np.roll(np.diag(np.asarray(vector).flatten()), 1, 1), -1, 0)
    return (skv - skv.T)


def transform_between_vectors(vector_a, vector_b):
    """
    Compute the transformation that aligns two vectors

    Parameters
    ----------
    vector_a: array_like
      The initial vector
    vector_b: array_like
      The goal vector

    Returns
    -------
    transform: array_like
      The transformation between `vector_a` a `vector_b`
    """
    newaxis = unit(vector_b)
    oldaxis = unit(vector_a)
    # Limits the value of `c` to be within the range C{[-1, 1]}
    c = np.clip(np.dot(oldaxis, newaxis), -1., 1.)
    angle = np.arccos(c)
    if np.isclose(c, -1.0) or np.allclose(newaxis, oldaxis):
        axis = perpendicular(newaxis)
    else:
        axis = unit(np.cross(oldaxis, newaxis))
    transform = br.axis_angle.to_transform(axis, angle)
    return transform


src/baldor/__init__.py
from . import axis_angle
from . import euler
from . import quaternion
from . import transform
from . import vector

# Utils
import numpy as np
X_AXIS = np.array([1., 0., 0.], dtype=np.float64)
Y_AXIS = np.array([0., 1., 0.], dtype=np.float64)
Z_AXIS = np.array([0., 0., 1.], dtype=np.float64)
_MAX_FLOAT = np.maximum_sctype(np.float)
# epsilon for testing whether a number is close to zero
_FLOAT_EPS = np.finfo(np.float).eps
_EPS = np.finfo(float).eps * 4.0
# axis sequences for Euler angles
_NEXT_AXIS = [1, 2, 0, 1]
# map axes strings to/from tuples of inner axis, parity, repetition, frame
_AXES2TUPLE = {
    'sxyz': (0, 0, 0, 0), 'sxyx': (0, 0, 1, 0), 'sxzy': (0, 1, 0, 0),
    'sxzx': (0, 1, 1, 0), 'syzx': (1, 0, 0, 0), 'syzy': (1, 0, 1, 0),
    'syxz': (1, 1, 0, 0), 'syxy': (1, 1, 1, 0), 'szxy': (2, 0, 0, 0),
    'szxz': (2, 0, 1, 0), 'szyx': (2, 1, 0, 0), 'szyz': (2, 1, 1, 0),
    'rzyx': (0, 0, 0, 1), 'rxyx': (0, 0, 1, 1), 'ryzx': (0, 1, 0, 1),
    'rxzx': (0, 1, 1, 1), 'rxzy': (1, 0, 0, 1), 'ryzy': (1, 0, 1, 1),
    'rzxy': (1, 1, 0, 1), 'ryxy': (1, 1, 1, 1), 'ryxz': (2, 0, 0, 1),
    'rzxz': (2, 0, 1, 1), 'rxyz': (2, 1, 0, 1), 'rzyz': (2, 1, 1, 1)}
_TUPLE2AXES = dict((v, k) for k, v in _AXES2TUPLE.items())


src/baldor/transform.py
#!/usr/bin/env python
"""
Homogeneous Transformation Matrices
"""
import math
import numpy as np
# Local modules
import baldor as br


def are_equal(T1, T2, rtol=1e-5, atol=1e-8):
    """
    Returns True if two homogeneous transformation are equal within a tolerance.

    Parameters
    ----------
    T1: array_like
      First input homogeneous transformation
    T2: array_like
      Second input homogeneous transformation
    rtol: float
      The relative tolerance parameter.
    atol: float
      The absolute tolerance parameter.

    Returns
    -------
    equal : bool
       True if `T1` and `T2` are `almost` equal, False otherwise

    See Also
    --------
    numpy.allclose: Contains the details about the tolerance parameters
    """
    M1 = np.array(T1, dtype=np.float64, copy=True)
    M1 /= M1[3, 3]
    M2 = np.array(T2, dtype=np.float64, copy=True)
    M2 /= M2[3, 3]
    return np.allclose(M1, M2, rtol, atol)


def between_axes(axis_a, axis_b):
    """
    Compute the transformation that aligns two vectors/axes.

    Parameters
    ----------
    axis_a: array_like
      The initial axis
    axis_b: array_like
      The goal axis

    Returns
    -------
    transform: array_like
      The transformation that transforms `axis_a` into `axis_b`
    """
    a_unit = br.vector.unit(axis_a)
    b_unit = br.vector.unit(axis_b)
    c = np.dot(a_unit, b_unit)
    angle = np.arccos(c)
    if np.isclose(c, -1.0) or np.allclose(a_unit, b_unit):
        axis = br.vector.perpendicular(b_unit)
    else:
        axis = br.vector.unit(np.cross(a_unit, b_unit))
    transform = br.axis_angle.to_transform(axis, angle)
    return transform


def inverse(transform):
    """
    Compute the inverse of an homogeneous transformation.

    .. note:: This function is more efficient than :obj:`numpy.linalg.inv` given
      the special properties of homogeneous transformations.

    Parameters
    ----------
    transform: array_like
      The input homogeneous transformation

    Returns
    -------
    inv: array_like
      The inverse of the input homogeneous transformation
    """
    R = transform[:3, :3].T
    p = transform[:3, 3]
    inv = np.eye(4)
    inv[:3, :3] = R
    inv[:3, 3] = np.dot(-R, p)
    return inv


def random(max_position=1.):
    """
    Generate a random homogeneous transformation.

    Parameters
    ----------
    max_position: float, optional
      Maximum value for the position components of the transformation

    Returns
    -------
    T: array_like
      The random homogeneous transformation

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> T = br.transform.random()
    >>> Tinv = br.transform.inverse(T)
    >>> np.allclose(np.dot(T, Tinv), np.eye(4))
    True
    """
    quat = br.quaternion.random()
    T = br.quaternion.to_transform(quat)
    T[:3, 3] = np.random.rand(3)*max_position
    return T


def to_axis_angle(transform):
    """
    Return rotation angle and axis from rotation matrix.

    Parameters
    ----------
    transform: array_like
      The input homogeneous transformation

    Returns
    -------
    axis: array_like
      axis around which rotation occurs
    angle: float
      angle of rotation
    point: array_like
      point around which the rotation is performed

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> axis = np.random.sample(3) - 0.5
    >>> angle = (np.random.sample(1) - 0.5) * (2*np.pi)
    >>> point = np.random.sample(3) - 0.5
    >>> T0 = br.axis_angle.to_transform(axis, angle, point)
    >>> axis, angle, point = br.transform.to_axis_angle(T0)
    >>> T1 = br.axis_angle.to_transform(axis, angle, point)
    >>> br.transform.are_equal(T0, T1)
    True
    """
    R = np.array(transform, dtype=np.float64, copy=False)
    R33 = R[:3, :3]
    # direction: unit eigenvector of R33 corresponding to eigenvalue of 1
    w, W = np.linalg.eig(R33.T)
    i = np.where(abs(np.real(w) - 1.0) < 1e-8)[0]
    if not len(i):
        raise ValueError("no unit eigenvector corresponding to eigenvalue 1")
    axis = np.real(W[:, i[-1]]).squeeze()
    # point: unit eigenvector of R corresponding to eigenvalue of 1
    w, Q = np.linalg.eig(R)
    i = np.where(abs(np.real(w) - 1.0) < 1e-8)[0]
    if not len(i):
        raise ValueError("no unit eigenvector corresponding to eigenvalue 1")
    point = np.real(Q[:, i[-1]]).squeeze()
    point /= point[3]
    # rotation angle depending on axis
    cosa = (np.trace(R33) - 1.0) / 2.0
    if abs(axis[2]) > 1e-8:
        sina = (R[1, 0] + (cosa-1.0)*axis[0]*axis[1]) / axis[2]
    elif abs(axis[1]) > 1e-8:
        sina = (R[0, 2] + (cosa-1.0)*axis[0]*axis[2]) / axis[1]
    else:
        sina = (R[2, 1] + (cosa-1.0)*axis[1]*axis[2]) / axis[0]
    angle = math.atan2(sina, cosa)
    return axis, angle, point


def to_dual_quaternion(transform):
    """
    Return quaternion from the rotation part of an homogeneous transformation.

    Parameters
    ----------
    transform: array_like
      Rotation matrix. It can be (3x3) or (4x4)
    isprecise: bool
      If True, the input transform is assumed to be a precise rotation matrix and
      a faster algorithm is used.

    Returns
    -------
    qr: array_like
      Quaternion in w, x, y z (real, then vector) for the rotation component
    qt: array_like
      Quaternion in w, x, y z (real, then vector) for the translation component

    Notes
    -----
    Some literature prefers to use :math:`q` for the rotation component and
    :math:`q'` for the translation component
    """
    def cot(x): return 1./np.tan(x)
    R = np.eye(4)
    R[:3, :3] = transform[:3, :3]
    l, theta, _ = to_axis_angle(R)
    t = transform[:3, 3]
    # Pitch d
    d = np.dot(l.reshape(1, 3), t.reshape(3, 1))
    # Point c
    c = 0.5*(t-d*l) + cot(theta/2.)*np.cross(l, t)
    # Moment vector
    m = np.cross(c, l)
    # Rotation quaternion
    qr = np.zeros(4)
    qr[0] = np.cos(theta/2.)
    qr[1:] = np.sin(theta/2.)*l
    # Translation quaternion
    qt = np.zeros(4)
    qt[0] = -(1/2.)*np.dot(qr[1:], t)
    qt[1:] = (1/2.)*(qr[0]*t + np.cross(t, qr[1:]))
    return qr, qt


def to_euler(transform, axes='sxyz'):
    """
    Return Euler angles from transformation matrix with the specified axis
    sequence.

    Parameters
    ----------
    transform: array_like
      Rotation matrix. It can be (3x3) or (4x4)
    axes: str, optional
      Axis specification; one of 24 axis sequences as string or encoded tuple

    Returns
    -------
    ai: float
      First rotation angle (according to axes).
    aj: float
      Second rotation angle (according to axes).
    ak: float
      Third rotation angle (according to axes).

    Notes
    -----
    Many Euler angle triplets can describe the same rotation matrix

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> T0 = br.euler.to_transform(1, 2, 3, 'syxz')
    >>> al, be, ga = br.transform.to_euler(T0, 'syxz')
    >>> T1 = br.euler.to_transform(al, be, ga, 'syxz')
    >>> np.allclose(T0, T1)
    True
    """
    try:
        firstaxis, parity, repetition, frame = br._AXES2TUPLE[axes.lower()]
    except (AttributeError, KeyError):
        br._TUPLE2AXES[axes]  # validation
        firstaxis, parity, repetition, frame = axes

    i = firstaxis
    j = br._NEXT_AXIS[i+parity]
    k = br._NEXT_AXIS[i-parity+1]

    M = np.array(transform, dtype=np.float64, copy=False)[:3, :3]
    if repetition:
        sy = math.sqrt(M[i, j]*M[i, j] + M[i, k]*M[i, k])
        if sy > br._EPS:
            ax = math.atan2(M[i, j],  M[i, k])
            ay = math.atan2(sy,      M[i, i])
            az = math.atan2(M[j, i], -M[k, i])
        else:
            ax = math.atan2(-M[j, k],  M[j, j])
            ay = math.atan2(sy,      M[i, i])
            az = 0.0
    else:
        cy = math.sqrt(M[i, i]*M[i, i] + M[j, i]*M[j, i])
        if cy > br._EPS:
            ax = math.atan2(M[k, j],  M[k, k])
            ay = math.atan2(-M[k, i],  cy)
            az = math.atan2(M[j, i],  M[i, i])
        else:
            ax = math.atan2(-M[j, k],  M[j, j])
            ay = math.atan2(-M[k, i],  cy)
            az = 0.0

    if parity:
        ax, ay, az = -ax, -ay, -az
    if frame:
        ax, az = az, ax
    return ax, ay, az


def to_quaternion(transform, isprecise=False):
    """
    Return quaternion from the rotation part of an homogeneous transformation.

    Parameters
    ----------
    transform: array_like
      Rotation matrix. It can be (3x3) or (4x4)
    isprecise: bool
      If True, the input transform is assumed to be a precise rotation matrix and
      a faster algorithm is used.

    Returns
    -------
    q: array_like
      Quaternion in w, x, y z (real, then vector) format

    Notes
    -----
    Quaternions :math:`w + ix + jy + kz` are represented as :math:`[w, x, y, z]`.

    Examples
    --------
    >>> import numpy as np
    >>> import baldor as br
    >>> q = br.transform.to_quaternion(np.identity(4), isprecise=True)
    >>> np.allclose(q, [1, 0, 0, 0])
    True
    >>> q = br.transform.to_quaternion(np.diag([1, -1, -1, 1]))
    >>> np.allclose(q, [0, 1, 0, 0]) or np.allclose(q, [0, -1, 0, 0])
    True
    >>> T = br.axis_angle.to_transform((1, 2, 3), 0.123)
    >>> q = br.transform.to_quaternion(T, True)
    >>> np.allclose(q, [0.9981095, 0.0164262, 0.0328524, 0.0492786])
    True
    """
    M = np.array(transform, dtype=np.float64, copy=False)[:4, :4]
    if isprecise:
        q = np.empty((4, ))
        t = np.trace(M)
        if t > M[3, 3]:
            q[0] = t
            q[3] = M[1, 0] - M[0, 1]
            q[2] = M[0, 2] - M[2, 0]
            q[1] = M[2, 1] - M[1, 2]
        else:
            i, j, k = 0, 1, 2
            if M[1, 1] > M[0, 0]:
                i, j, k = 1, 2, 0
            if M[2, 2] > M[i, i]:
                i, j, k = 2, 0, 1
            t = M[i, i] - (M[j, j] + M[k, k]) + M[3, 3]
            q[i] = t
            q[j] = M[i, j] + M[j, i]
            q[k] = M[k, i] + M[i, k]
            q[3] = M[k, j] - M[j, k]
            q = q[[3, 0, 1, 2]]
        q *= 0.5 / math.sqrt(t * M[3, 3])
    else:
        m00 = M[0, 0]
        m01 = M[0, 1]
        m02 = M[0, 2]
        m10 = M[1, 0]
        m11 = M[1, 1]
        m12 = M[1, 2]
        m20 = M[2, 0]
        m21 = M[2, 1]
        m22 = M[2, 2]
        # symmetric matrix K
        K = np.array([[m00-m11-m22, 0.0,         0.0,         0.0],
                      [m01+m10,     m11-m00-m22, 0.0,         0.0],
                      [m02+m20,     m12+m21,     m22-m00-m11, 0.0],
                      [m21-m12,     m02-m20,     m10-m01,     m00+m11+m22]])
        K /= 3.0
        # quaternion is eigenvector of K that corresponds to largest eigenvalue
        w, V = np.linalg.eigh(K)
        q = V[[3, 0, 1, 2], np.argmax(w)]
    if q[0] < 0.0:
        np.negative(q, q)
    return q
